File: test/modulebase/TestThreeCenterIntegral.py
import pulsar as psr
import numpy as np

Mat1 = np.array([1.1,2.2,3.3])

class Test3CInt(psr.ThreeCenterIntegral):
    def __init__(self,id):
        super(Test3CInt,self).__init__(id)

    def initialize_(self,deriv,wf,bs1,bs2,bs3):
        return

    def calculate_(self,shell1,shell2,shell3):
        return Mat1

def run_test():
    tester=psr.PyTester("Testing ThreeCenterIntegral Python Bindings")
    mm=psr.ModuleManager()
    mm.load_lambda_module(Test3CInt,"IntBuilder","test_builder")
    builder=mm.get_module("test_builder",0)
    deriv=1
    wf=psr.Wavefunction()
    bs=psr.BasisSet(1,3,3,3)
    tester.test_call("Can call initialize",True,builder.initialize,deriv,wf,bs,bs,bs)
    temp = builder.calculate(deriv,deriv,deriv).tolist()
    tester.test_double_vector("Can call calculate",Mat1,temp)

    tester.print_results()
    return tester.nfailed()

File: test/modulebase/TestThreeCenterIntegral.cpp
#include <pulsar/testing/CppTester.hpp>
#include <pulsar/modulemanager/ModuleManager.hpp>
#include <pulsar/modulebase/ThreeCenterIntegral.hpp>

using namespace std;
using namespace pulsar;

std::vector<double> corr_answer={1.1,2.2,3.3};

class Test3CInt:public ThreeCenterIntegral{
public:
    Test3CInt(ID_t id):ThreeCenterIntegral(id){}
private:
    void initialize_(unsigned int,
                     const Wavefunction &,
                     const BasisSet &,
                     const BasisSet &,
                     const BasisSet &)
    {

    }

    HashType my_hash_(unsigned int,
                      const Wavefunction &,
                      const BasisSet &,
                      const BasisSet &,
                      const BasisSet &){
        return "";
    }

    const double* calculate_(uint64_t, uint64_t, uint64_t)
    {
        return corr_answer.data();
    }
};

TEST_SIMPLE(TestThreeCenterIntegral){
    CppTester tester("Testing ThreeCenterIntegral base class");

    auto mm=make_shared<ModuleManager>();
    mm->load_lambda_module<Test3CInt>("ThreeCenterIntegral","test_module");
    auto builder=mm->get_module<Test3CInt>("test_module",0);

    unsigned int deriv=1;
    Wavefunction wf;
    BasisSet bs;

    ThreeCenterIntegral* ints=&(*builder);
    tester.test_member_call("Can_call initialize",true,
                            &ThreeCenterIntegral::initialize,ints,deriv,wf,bs,bs,bs);

    const double* corr=corr_answer.data();

    tester.test_member_return("Can call calculate",true,corr,
        &ThreeCenterIntegral::calculate,ints,deriv,deriv,deriv);
    tester.print_results();
    return tester.nfailed();
}
